libkbfs/keybase_daemon_local.go

```diff
@@ -236,6 +250,139 @@ func (k *KeybaseDaemonLocal) Identify(
 	return k.Resolve(ctx, assertion)
 }
 
+// ResolveIdentifyImplicitTeam implements the KeybaseService interface
+// for KeybaseDaemonLocal.
+func (k *KeybaseDaemonLocal) ResolveIdentifyImplicitTeam(
+	ctx context.Context, assertions, suffix string, tlfType tlf.Type,
+	doIdentifies bool, reason string) (ImplicitTeamInfo, error) {
+	if err := checkContext(ctx); err != nil {
+		return ImplicitTeamInfo{}, err
+	}
+
+	if tlfType != tlf.Private && tlfType != tlf.Public {
+		return ImplicitTeamInfo{}, fmt.Errorf(
+			"Invalid implicit team TLF type: %s", tlfType)
+	}
+
+	k.lock.Lock()
+	defer k.lock.Unlock()
+
+	// Canonicalize the name.
+	writerNames, readerNames, _, err :=
+		splitAndNormalizeTLFName(assertions, tlfType)
+	if err != nil {
+		return ImplicitTeamInfo{}, err
+	}
+	var writers, readers []libkb.NormalizedUsername
+	var unresolvedWriters, unresolvedReaders []keybase1.SocialAssertion
+	resolve := func(
+		name string, r []libkb.NormalizedUsername,
+		ur []keybase1.SocialAssertion) (
+		[]libkb.NormalizedUsername, []keybase1.SocialAssertion, error) {
+		id, err := k.assertionToIDLocked(ctx, name)
+		if err == nil {
+			u, err := k.localUsers.getLocalUser(id.AsUserOrBust())
+			if err != nil {
+				return nil, nil, err
+			}
+			r = append(r, u.Name)
+		} else {
+			a, ok := externals.NormalizeSocialAssertion(name)
+			if !ok {
+				return nil, nil, fmt.Errorf("Bad assertion: %s", name)
+			}
+			ur = append(ur, a)
+		}
+		return r, ur, nil
+	}
+	for _, w := range writerNames {
+		writers, unresolvedWriters, err = resolve(w, writers, unresolvedWriters)
+		if err != nil {
+			return ImplicitTeamInfo{}, err
+		}
+	}
+	for _, r := range readerNames {
+		readers, unresolvedReaders, err = resolve(r, readers, unresolvedReaders)
+		if err != nil {
+			return ImplicitTeamInfo{}, err
+		}
+	}
+
+	var extensions []tlf.HandleExtension
+	if len(suffix) != 0 {
+		extensions, err = tlf.ParseHandleExtensionSuffix(suffix)
+		if err != nil {
+			return ImplicitTeamInfo{}, err
+		}
+	}
+	name := tlf.MakeCanonicalName(
+		writers, unresolvedWriters, readers, unresolvedReaders, extensions)
+
+	key := fmt.Sprintf("%s:%s", tlfType.String(), name)
+	tid, ok := k.implicitAsserts[key]
+	if ok {
+		return k.localImplicitTeams[tid], nil
+	}
+
+	if !doIdentifies {
+		return ImplicitTeamInfo{}, NoSuchTeamError{assertions}
+	}
+
+	// Need to make the team info as well, so get the list of user
+	// names and resolve them.
+	asUserName := libkb.NormalizedUsername(name)
+	teams := makeLocalTeams(
+		[]libkb.NormalizedUsername{asUserName}, len(k.localTeams), tlfType)
+	info := teams[0]
+	info.Writers = make(map[keybase1.UID]bool, len(writerNames))
+	for _, w := range writerNames {
+		id, err := k.assertionToIDLocked(ctx, w)
```

Comment:
We've already done this in the `resolve` function above, and an error in resolving to either a name or social assertion will have caused this whole function to fail. As a bonus, we also already have the social assertions in `unresolved{Readers,Writers}`. So maybe this loop can iterate over `range resolvedWriters` instead of `writerNames`. Not sure how you'd like to handle the social assertions.